### PR TITLE
Added config utils

### PR DESCRIPTION
**Pull request by samyfodil**

### CODE REVIEW

File: cli/app/gen_config.go [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+		if err != nil {
+			return "", err
+		}
+
+		err = os.WriteFile(path.Join(configRoot, "keys", "dv_private.pem"), priv, 0440)
+		if err != nil {
+			return "", err
+		}
+
+		err = os.WriteFile(path.Join(configRoot, "keys", "dv_public.pem"), pub, 0440)
```
Comment (tafseer-khan) [inline cli/app/gen_config.go:110]: same

File: cli/app/gen_config.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+		}
+	}
+
+	if ctx.Bool("dv-keys") {
+		priv, pub, err := generateDVKeys()
+		if err != nil {
+			return "", err
+		}
+
+		err = os.WriteFile(path.Join(configRoot, "keys", "dv_private.pem"), priv, 0440)
```
Comment (tafseer-khan) [inline cli/app/gen_config.go:105]: same

File: cli/app/gen_config.go [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+	}
+
+	f, err := os.Create(configPath)
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+
+	yamlEnc := yaml.NewEncoder(f)
+	err = yamlEnc.Encode(configStruct)
```
Comment (tafseer-khan) [inline cli/app/gen_config.go:123]: same

File: cli/app/gen_config.go [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+	if err != nil {
+		return "", err
+	}
+
+	return nodeID, nil
+}
+
+func getProtocols(s string) []string {
+	protos := make(map[string]bool)
```
Comment (tafseer-khan) [inline cli/app/gen_config.go:132]: can make with fixed size

File: cli/app/gen_config.go [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+	for _, p := range commonSpecs.Protocols {
+		protos[p] = false
+	}
+	for _, p := range strings.Split(s, ",") {
```
Comment (tafseer-khan) [inline cli/app/gen_config.go:136]: new line ebefore

File: cli/app/gen_config.go [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+		if _, ok := protos[p]; ok {
+			protos[p] = true
+		}
+	}
+	ret := make([]string, 0, len(commonSpecs.Protocols))
```
Comment (tafseer-khan) [inline cli/app/gen_config.go:141]: new line before

File: cli/app/gen_config.go [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+	for p, on := range protos {
+		if on {
+			ret = append(ret, p)
+		}
+	}
+	return ret
```
Comment (tafseer-khan) [inline cli/app/gen_config.go:147]: new line

File: cli/app/gen_config.go [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+	if err != nil {
+		return "", fmt.Errorf("rand read failed with: %w", err)
+	}
+
+	return "/key/swarm/psk/1.0.0//base16/" + hex.EncodeToString(key), nil
+}
+
+func generateNodeKeyAndID() (string, string, error) {
+	key := keypair.New()
+
```
Comment (tafseer-khan) [inline cli/app/gen_config.go:162]: remove \n

File: cli/app/gen_config.go [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+		return "", "", fmt.Errorf("id from private key failed with %w", err)
+	}
+
+	return id.String(), base64.StdEncoding.EncodeToString(keyData), nil
+}
+
+func generateDVKeys() ([]byte, []byte, error) {
+	priv, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return nil, nil, fmt.Errorf("generate ecdsa key failed with %w", err)
```
Comment (tafseer-khan) [inline cli/app/gen_config.go:179]: with: %w

File: cli/app/gen_config.go [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+
+func pemEncodePrivKey(priv *ecdsa.PrivateKey) ([]byte, error) {
+	privateKey, err := x509.MarshalECPrivateKey(priv)
+	if err != nil {
+		return nil, fmt.Errorf("marshal private key failed with %w", err)
+	}
+
+	privateKeyPEM := &pem.Block{Type: "EC PRIVATE KEY", Bytes: privateKey}
+	var privateBuf bytes.Buffer
+	err = pem.Encode(&privateBuf, privateKeyPEM)
```
Comment (tafseer-khan) [inline cli/app/gen_config.go:203]: 1 line

File: cli/app/gen_config.go [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+func pemEncodePubKey(priv *ecdsa.PrivateKey) ([]byte, error) {
+	// Generate public key and write to a file
+	publicKey, err := x509.MarshalPKIXPublicKey(&priv.PublicKey)
+	if err != nil {
+		return nil, fmt.Errorf("marshal public key failed with %w", err)
+	}
+
+	publicKeyPEM := &pem.Block{Type: "PUBLIC KEY", Bytes: publicKey}
+	var public bytes.Buffer
+	err = pem.Encode(&public, publicKeyPEM)
```
Comment (tafseer-khan) [inline cli/app/gen_config.go:220]: 1 line

File: cli/app/parse_config.go [hunk 12]
```diff
@@ -17,16 +17,24 @@ import (
...
 	root := ctx.Path("root")
 
 	if !filepath.IsAbs(root) {
 		return nil, nil, fmt.Errorf("root folder `%s` is not absolute", root)
 	}
 
 	configRoot := root + "/config"
-	configPath := path.Join(configRoot, ctx.String("shape")+".yaml")
+
+	var configPath string
```
Comment (tafseer-khan) [inline cli/app/parse_config.go:32]: configPath := ctx.Path("path")

if len(configPath ) < 1{
configPath = path.Join ...
}

File: cli/app/parse_config.go [hunk 13]
```diff
@@ -84,10 +92,7 @@ func parseSourceConfig(ctx *cli.Context) (*config.Protocol, *config.Source, erro
 		return nil, nil, err
 	}
 
-	protocol.DomainValidation.PrivateKey, protocol.DomainValidation.PublicKey, err = parseValidationKey(
-		src.Domains.Key.Private,
-		src.Domains.Key.Public,
-	)
+	protocol.DomainValidation, err = parseValidationKey(&src.Domains.Key)
```
Comment (tafseer-khan) [inline cli/app/parse_config.go:95]: one line

File: cli/app/parse_config.go [hunk 14]
```diff
@@ -108,28 +113,28 @@ func parseSwarmKey(filepath string) (pnet.PSK, error) {
...
 }
 
-func parseValidationKey(privateKeyPath, publicKeyPath string) ([]byte, []byte, error) {
+func parseValidationKey(key *config.DVKey) (config.DomainValidation, error) {
 	// Private Key
-	privateKey, err := os.ReadFile(privateKeyPath)
+	privateKey, err := os.ReadFile(key.Private)
 	if err != nil {
-		return nil, nil, fmt.Errorf("reading private key `%s` failed with: %s", privateKeyPath, err)
+		return config.DomainValidation{}, fmt.Errorf("reading private key `%s` failed with: %s", key.Private, err)
```
Comment (tafseer-khan) [inline cli/app/parse_config.go:120]: %w

**Review by tafseer-khan — CHANGES_REQUESTED** [review]

File: cli/app/config_cmd.go [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+
+func configCommand() *cli.Command {
+	return &cli.Command{
+		Name:        "config",
+		Aliases:     []string{"cnf", "conf"},
+		Description: "configuration utils",
+		Subcommands: []*cli.Command{
+			{
+				Name:    "validate",
+				Aliases: []string{"check", "ok", "ok?", "valid?"},
```
[the remark this answers is not in the document]
Comment (samyfodil) [reply cli/app/config_cmd.go:18]: yes

File: cli/app/config_cmd.go [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+					},
+					&cli.BoolFlag{
+						Name:    "dv-keys",
+						Aliases: []string{"dv"},
+					},
+				},
+				Action: func(ctx *cli.Context) error {
+					id, err := generateSourceConfig(ctx)
+					if id != "" {
+						fmt.Println("ID:", id)
```
[the remark this answers is not in the document]
Comment (samyfodil) [reply cli/app/config_cmd.go:83]: can you make the change

File: cli/app/gen_config.go [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+
+func generateSourceConfig(ctx *cli.Context) (string, error) {
+	root := ctx.Path("root")
+
+	if !filepath.IsAbs(root) {
+		return "", fmt.Errorf("root folder `%s` is not absolute", root)
+	}
+
+	configRoot := root + "/config"
+
```
[the remark this answers is not in the document]
Comment (samyfodil) [reply cli/app/gen_config.go:40]: do it

File: cli/app/gen_config.go [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+		},
+	}
+
+	if ctx.Bool("swarm-key") {
+		swarmkey, err := generateSwarmKey()
+		if err != nil {
+			return "", err
+		}
+
+		err = os.WriteFile(path.Join(configRoot, "keys", "swarm.key"), []byte(swarmkey), 0440)
```
[the remark this answers is not in the document]
Comment (samyfodil) [reply cli/app/gen_config.go:93]: sure

File: cli/app/gen_config.go [hunk 19]
```diff
@@ -0,0 +1,226 @@
...
+		if on {
+			ret = append(ret, p)
+		}
+	}
+	return ret
+}
+
+func generateSwarmKey() (string, error) {
+	key := make([]byte, 32)
+	_, err := rand.Read(key)
```
[the remark this answers is not in the document]
Comment (samyfodil) [reply cli/app/gen_config.go:152]: do it

File: cli/app/gen_config.go (hunk 10 above)
Comment (samyfodil) [reply cli/app/gen_config.go:203]: do it

File: cli/app/gen_config.go (hunk 11 above)
Comment (samyfodil) [reply cli/app/gen_config.go:220]: do it

File: cli/app/parse_config.go (hunk 12 above)
Comment (samyfodil) [reply cli/app/parse_config.go:32]: don't understand

File: cli/app/parse_config.go (hunk 13 above)
Comment (samyfodil) [reply cli/app/parse_config.go:95]: do it

File: cli/app/parse_config.go (hunk 14 above)
Comment (samyfodil) [reply cli/app/parse_config.go:120]: do it

File: cli/app/config_cmd.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (samyfodil) [reply cli/app/config_cmd.go:18]: to be kept

File: cli/app/parse_config.go (hunk 12 above)
Comment (tafseer-khan) [reply cli/app/parse_config.go:32]: No need to do if else

**Pull request merged**